### ISSUE TITLE
#Store Function

### ISSUE THREAD

**Issue by DomNidy** [cabaletta/baritone#605]
Im sure someone has already recommended this but I think a storage function would be excellent for baritone. Whenever your inventory gets full from mining or farming the bot would automatically store all added items to nearby chests/marked chests.

The store function would work similar to the mine function and would have a configurable list of items to store. ( store cobblestone diamond_ore iron_ore coal)
As for configuring where to store the function could work with the #goto function and goto the nearest chest, and there could also be another function to mark specific chests to store items in. #markchest (name of storage) (items to store here)

This would help baritone out by increasing automation time for mining & farming due to limitless storage.

**Comment by ZeroMemes** [comment]
@leijurv this could be done by having a higher priority process that takes control if you have the setting on and the inventory is full then activate and path accordingly?

**Comment by xaengceilbiths** [comment]
I'd use rightclick() in GetToBlockProcess to open a chest and Minecraft.getMinecraft().displayGuiScreen(null) to close a chest, am I right?
How many ticks should baritone REQUEST_PAUSE between open and moveitem and close?

**Comment by matthewfcarlson** [comment]
@dominikaaaa and @ZeroMemes is this something that baritone wants as a feature? I'm working on this for my personal usage and I'd be happy to polish it up to where I'd be comfortable pushing it back up.

**Comment by dominikaaaa** [comment]
I think so, this has been heavily requested and has a few duplicates. Go for it if you'd like.

**Comment by matthewfcarlson** [comment]
I'll put in some work on it in the evenings :) Let you know if I've got anything good. Here's the roadmap
[] Add new option to set the items you want to store (this would default to what you're mining)
[] Store items from shulkers in inventory if full
[] Pull items from shulkers in inventory if needed
[] Go put items in chests (leverage chests command?)
[] Pull items from chests if needed

**Comment by dominikaaaa** [comment]
That looks good so far, maybe have the second off by default and the rest could maybe be on by default

**Comment by matthewfcarlson** [comment]
Feel free to assign this one to me

**Comment by matthewfcarlson** [comment]
Yeah- those are good defaults

**Comment by matthewfcarlson** [comment]
Update: So I had been hoping I could just write the NBT data but it looks like the server doesn't like it and just overwrites it. So right now, it finds a spot to place it, places it, opens it, tries to place things inside (inventory management gets a little wonky when it's full), mines it, then picks up the shulker box. I might break out the retrieving functionality into a separate PR/issue.

**Comment by matthewfcarlson** [comment]
Is there functionality already for this in impact? I don't want to be duplicating work

**Comment by dominikaaaa** [comment]
I don't believe so, and impact is closed source anyways. People still use baritone standalone or in their own mods without impact.

**Comment by michaelzangl** [comment]
For minebot, I made good experience with storing into normal chests with signs / item frames attached. The algorithm is simple:
* Scan chests around the player
* Find the one with a matching item in a item frame or a sign with the given item name on it
* Walk near that chest
* Open it
* Place as many items in there as possible

Repeat if the inventory contains more items (chests may be full and client only knows after opening it)

The advantage I saw in putting item frames / signs on the chests is that the items will be sorted and that it is pretty easy to configure.

The same strategy can be used for getting items, skipping empty chests.

**Comment by dominikaaaa** [comment]
I saw your 1.15.2 branch the other day and I can't believe you're working on Minebot again lol @michaelzangl

**Comment by michaelzangl** [comment]
@dominikaaaa I won't reactivate all functionality. Minebot is my personal playground project. I currently plan on using it as testing ground for kotlin script integration, to replace the deprecated and old javascript engine that Java comes with.

**Comment by beartravels** [comment]
I just wanted to leave a comment to how cool this would be in baritone, i build large schematics and want to schematic my red stone bases for easy reproduction but resupplying the bot with only the blocks in my  inventory is a pain in the backside. so good luck and i hope all is going well

**Comment by Teals53** [comment]
any news about this ?

**Comment by l1ving** [comment]
The issue is still open

**Comment by Teals53** [comment]
so nothing

**Comment by l1ving** [comment]
So this issue is still open...

**Comment by l1ving** [comment]
#1706 exists if you want to try it, but it's not merged

**Comment by Teals53** [comment]
i dont have any software for compile this version also im on mobile internet atm can u help me ? @l1ving also thx for fast reply

**Comment by OmegaKing** [comment]
Has there been any progress for this? If so when do you think it will be pushed out?

**Comment by l1ving** [comment]
> If so when do you think it will be pushed out?

![image](https://user-images.githubusercontent.com/17222512/99118916-5f722300-25c6-11eb-8862-d21215526ec0.png)

# :joy_cat:

**Comment by filiphak1** [comment]
You guys are working on function that stores items in chests/shulkers right? It would be really great cuz I want to leave my character to mine resources whole night and in the meantime store those mined items in chests :D

**Comment by matthewfcarlson** [comment]
I think that's the plan. It's a huge lift. I've been giving it some thought and I think here's all the things that needs to happen if this is going to get pulled off. Feel free to correct or add to this list

- Augment the task system support proper interrupting and reordering
- Fix the mine task to allow it to be interrupted
- Finish the chest storage location system and add a command to remember chests
- add navigate to a remembered chest functionality
- add helpers to manage storage and retrieval
- add system that detects if inventory is full while mining and kicks off storage process
- add shulkers? (it was way harder than I'd like and got ugly) 


I might dive back into this during the Thanksgiving holiday here in the US, but this is a huge amount of work. If you were to pay someone to do this I'd expect a $10-20k quote.

**Comment by l1ving** [comment]
It's definitely a lot of work. Take your time. 

> add shulkers? (it was way harder than I'd like and got ugly)

This (https://github.com/kami-blue/client/blob/e80c719539a3005093c83ce1d6c120b5948ab4d3/src/main/java/me/zeroeightsix/kami/module/modules/misc/AutoObsidian.kt#L90) is using a Baritone process to manage everything, but runs on it's own and is compatible with Baritone's existing processes via pauses

**Comment by jacklollz2** [comment]
Your chests/shulkers could be defined using the sel command to create a region of depositable chests/shulkers.

**Comment by pencildoor** [comment]
I have an idea: if putting things in chests is so difficult, why not just drop the items into a storage system we build? We just walk over to our storage system or use a command like /home for various multiplayer servers and drop the items onto the ground where a hopper can pick them up and put them into our storage system. It's simple to implement and I, for one, would be very satisfied if baritone had only this functionality.

**Comment by joevento** [comment]
What is the current status of this feature? This would be such a cool feature. Really looking forward to using it once its done, if it isn't done already.

**Comment by InvisibleCatA1** [comment]
Something like this would be nice for autobuild incase you did not have room in your inventory for all the required items.

**Comment by DereC4** [comment]
Would be nice

**Comment by Moon-2006** [comment]
Four years old lol. Anyone have any updates on this?

**Comment by matthewfcarlson** [comment]
I had a little prototype that worked sometimes but found moving things in and out of inventories to be unreliable. It would try to drop off a load of coal and then start walking away, only to walk back because the inventory was still full. I think the idea of just dropping things on the ground is a great one, but my goal with this was to build impossibly large structures in survival mode. It's pretty low on my priorities, but I would like to get back to someday.

**Comment by Gamekind135** [comment]
i startet with baritone like a month ago, and ive seen videos from like 2bt2 where they automaticly get an shulker and get echests and then mine them for obbi, so i see this is already an i guess old ticket and are here already stuff like auto obsidian or auto fill in shulker while mining or farming or thad? Ive Loved to know this

**Comment by rycbar0** [comment]
while this people you have seen probably uses a client that contains baritone, base baritone as you get it from us does not have this features. they are added by the developer of the client by creating macros or custom baritone versions. to see what base baritone can do i suggest to read the command list in [usage](https://github.com/cabaletta/baritone/blob/master/USAGE.md) or consult the ingame help with `#help` which also provides you with a list of all commands, what they do and how to use them.

**Comment by Gamekind135** [comment]
i know these commands but how can i mod the baritone so it "Farm Obsidian" trough echest store items away afer minig and the only thing i want is this with fabric 1.19.2

**Comment by rycbar0** [comment]
you get the source code from here, modify it, compile it and put the compiled jar file in your mods folder.

**Comment by Moon-2006** [comment]
I know its possible to do, Lejurv almost certainly used a modified version of baritone to manage Gringotts, and the bots on 2b2t that automatically deliver resources might use something similar as well, but its none of it is public. Maybe someone will release code for it eventually.

**Comment by thaqalayn** [comment]
@leijurv can we make this happen if this hits 5 likes 🥺

**Comment by Nickyharbr** [comment]
anyone got this working

**Comment by WolfSurwail** [comment]
> Есть ли уже функциональность для этого? Я не хочу дублировать работу

after 4 years you were able to do this?

**Comment by TropicLegend** [comment]
Has anybody got this to work?
Sounds like an absolutely awesome feature to implement tbh.

**Comment by REALERvolker1** [comment]
I don't blame them for deciding not to implement this feature. It is very complex. You have to have the bot detect shulkers in your inventory, set a goal to place those down, automatically open the shulkers, deposit the items, then set another goal to mine the shulkers, then continue mining. Any slight variations or constraints in this process could easily and completely break it.
